src/adaptation/icap/Xaction.cc

```diff
@@ -442,6 +442,7 @@ void Adaptation::Icap::Xaction::noteCommRead(const CommIoCbParams &io)
 
     CommIoCbParams rd(this); // will be expanded with ReadNow results
     rd.conn = io.conn;
+    rd.size = SQUID_TCP_SO_RCVBUF - readBuf.length();
```

Comment:
> when this buffer contains `SQUID_TCP_SO_RCVBUF` bytes, the behaviour you are trying to avoid may occur despite this change.

Existing `Must(readBuf.length() < SQUID_TCP_SO_RCVBUF)` code executed a few lines earlier makes the precondition in the above statement false -- the buffer cannot contain `SQUID_TCP_SO_RCVBUF` bytes at this point.

> I believe the fix you actually want is to use `SBuf::reserve` to set the buffer limits properly

We have considered that approach and ruled it out: SBuf::reserve() is not the right solution for the problem targeted by this PR because SBuf::reserve() must not _shrink_ existing SBuf capacity. In modern C++, both std::vector::reserve() and std::string::reserve() do not shrink the underlying storage. C++ folks even explicitly [prohibited](https://en.cppreference.com/w/cpp/string/basic_string/reserve) such shrinking for strings in C++20!

We could consider adding a dedicated buffer shrinking method, but since we already have a documented API for avoiding excessive reads (without buffer shrinking), and lack performance data to justify shrinking overheads, we should use that rd.size API to fix the problem.